### PR TITLE
Repair constants using symbolic constructors

### PR DESCRIPTION
**Pull request by ajreynol**
This generalizes the repair constant utility of sygus so that the task of enumerating templates is done at the grammar level. This makes it so that we don't prune instantiations of repairable templates.

I am disabling the option by default since it will need further testing.

This fixes the timeout in regress2.

### CODE REVIEW

File: src/theory/datatypes/datatypes_sygus.cpp [hunk 1]
```diff
@@ -483,10 +484,25 @@ Node SygusSymBreakNew::getSimpleSymBreakPred( TypeNode tn, int tindex, unsigned
             for( unsigned i=0; i<deq_child[0].size(); i++ ){
               unsigned c1 = deq_child[0][i];
               unsigned c2 = deq_child[1][i];
-              if( children[c1].getType()==children[c2].getType() ){
-                if( !children[c1].getType().getCardinality().isOne() ){
-                  sbp_conj.push_back( children[c1].eqNode( children[c2] ).negate() );
+              TypeNode tnc = children[c1].getType();
+              if (tnc == children[c2].getType()
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_sygus.cpp]: doc to describe what this block is doing

File: src/theory/datatypes/datatypes_sygus.cpp [hunk 2]
```diff
@@ -558,6 +581,45 @@ Node SygusSymBreakNew::getSimpleSymBreakPred( TypeNode tn, int tindex, unsigned
           }else{
             // defined function?
           }
+          // explicitly handle "any constant" constructors
+          // if this type admits any constant, then at least one my children
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_sygus.cpp]: fix comment

File: src/theory/datatypes/datatypes_sygus.cpp [hunk 3]
```diff
@@ -558,6 +581,45 @@ Node SygusSymBreakNew::getSimpleSymBreakPred( TypeNode tn, int tindex, unsigned
...
+            for (unsigned j = 0; j < dt_index_nargs; j++)
+            {
+              TypeNode ctn = TypeNode::fromType(dt[tindex].getArgType(j));
+              int ctn_ac = d_tds->getAnyConstantConsNum(ctn);
+              if (ctn_ac == -1)
+              {
+                success = false;
+                break;
+              }
+              else
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_sygus.cpp]: no-else-after-break

File: src/theory/quantifiers/sygus/cegis.h [hunk 4]
```diff
@@ -164,6 +164,12 @@ class Cegis : public SygusModule
    * added as refinement lemmas.
    */
   std::unordered_set<unsigned> d_cegis_sample_refine;
+
+  //---------------------------------for sygus repair
+  /** are we using grammar-based repair? */
+  bool d_using_gr_repair;
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/cegis.h]: clarify what using "grammar-based repair" entails

File: src/theory/quantifiers/sygus/cegis.cpp [hunk 5]
```diff
@@ -156,6 +173,50 @@ bool Cegis::constructCandidates(const std::vector<Node>& enums,
...
+    // if the solution contains a subterm that must be repaired
+    if (mustRepair)
+    {
+      std::vector<Node> fail_cvs = enum_values;
+      Assert(candidates.size() == fail_cvs.size());
+      if (src->repairSolution(candidates, fail_cvs, candidate_values))
+      {
+        return true;
+      }
+      else
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/cegis.cpp]: no-else-after-return

File: src/theory/quantifiers/sygus/sygus_repair_const.h [hunk 6]
```diff
@@ -33,9 +33,9 @@ class CegConjecture;
  * This module is used to repair portions of candidate solutions. In particular,
  * given a synthesis conjecture:
  *   exists f. forall x. P( f, x )
- * and a candidate solution f = \x. t[x,c] where c are constants, this function
- * checks whether there exists a term of the form \x. t[x,c'] for some constants
- * c' such that:
+ * and a candidate solution f = \x. t[x,r] where c are repairable terms, this
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/sygus_repair_const.h]: "where c" --> "where r"

File: src/theory/quantifiers/sygus/sygus_grammar_norm.h [hunk 7]
```diff
@@ -36,6 +36,12 @@ namespace quantifiers {
 
 class SygusGrammarNorm;
 
+/** Attribute true for variables that represent any constant */
+struct SygusAnyConstAttributeId
+{
+};
+typedef expr::Attribute<SygusAnyConstAttributeId, bool> SygusAnyConstAttribute;
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/quantifiers/sygus/sygus_grammar_norm.h]: the typedef is consistent with the other attribute definitions.

**Pull request merged**